adam-core/src/main/scala/org/bdgenomics/adam/rdd/settheory/SetTheory.scala

```diff
@@ -0,0 +1,346 @@
+package org.bdgenomics.adam.rdd.settheory
+
+import org.apache.spark.rdd.RDD
+import org.bdgenomics.adam.models.ReferenceRegion
+import scala.collection.mutable.ListBuffer
+
+private[settheory] sealed abstract class SetTheory extends Serializable {
+
+  protected val threshold: Long
+  protected val optPartitionMap: Option[Array[Option[(ReferenceRegion, ReferenceRegion)]]]
+
+  /**
+   * The condition that should be met in order for the primitive to be
+   * computed.
+   *
+   * @param firstRegion The region to test against.
+   * @param secondRegion The region to test.
+   * @param distanceThreshold The threshold for the primitive.
+   * @return True if the threshold requirement is met.
+   *         False if the threshold requirement is not met.
+   */
+  protected def condition(firstRegion: ReferenceRegion,
+                          secondRegion: ReferenceRegion,
+                          distanceThreshold: Long = 0L): Boolean
+}
+
+private[settheory] trait SetTheoryPrimitive extends SetTheory {
+
+  /**
+   * Gets the partition bounds from a ReferenceRegion keyed Iterator
+   *
+   * @param iter The data on a given partition. ReferenceRegion keyed
+   * @return The bounds of the ReferenceRegions on that partition, in an Iterator
+   */
+  protected def getRegionBoundsFromPartition[X](
+    iter: Iterator[(ReferenceRegion, X)]): Iterator[Option[(ReferenceRegion, ReferenceRegion)]] = {
+
+    if (iter.isEmpty) {
+      // This means that there is no data on the partition, so we have no bounds
+      Iterator(None)
+    } else {
+      val firstRegion = iter.next
+      val lastRegion =
+        if (iter.hasNext) {
+          // we have to make sure we get the full bounds of this partition, this
+          // includes any extremely long regions. we include the firstRegion for
+          // the case that the first region is extremely long
+          (iter ++ Iterator(firstRegion)).maxBy(f => (f._1.referenceName, f._1.end, f._1.start))
+          // only one record on this partition, so this is the extent of the bounds
+        } else {
+          firstRegion
+        }
+      Iterator(Some((firstRegion._1, lastRegion._1)))
+    }
+  }
+}
+
+/**
+ * The parent class for all inter-collection set theory operations.
+ *
+ * @tparam T The left side row data.
+ * @tparam U The right side row data.
+ * @tparam RT The return type for the left side row data.
+ * @tparam RU The return type for the right side row data.
+ */
+private[settheory] abstract class SetTheoryBetweenCollections[T, U, RT, RU] extends SetTheory {
```

Comment:
OOC, is there a reason to prefer an abstract class to a trait here? I know that `ShuffleRegionJoin` is an abstract class, but I was looking through the history and couldn't figure out what was the impetus for the change. CCing in @ryan-williams who made the change in https://github.com/bigdatagenomics/adam/commit/9ff0fa5ec256bf057d29d5c2613e424ae7fdde01.